core/internal/runhistoryreader/parquet/iterator/record_iterator.go

```diff
@@ -0,0 +1,187 @@
+package iterator
+
+import (
+	"fmt"
+
+	"github.com/apache/arrow-go/v18/arrow"
+)
+
+type keyIteratorPair struct {
+	Key      string
+	Iterator columnIterator
+}
+
+// recordIterator iterates over columnIterators
+// to read columns in the given arrow.RecordBatch in chunks.
+type recordIterator struct {
+	record           arrow.RecordBatch
+	columns          []keyIteratorPair
+	resultCols       []keyIteratorPair
+	resultsWhitelist map[string]struct{}
+
+	requireAllValues bool
+	indexKey         string
+	indexCol         columnIterator
+	validIndex       func(columnIterator) bool
+	value            KeyValueList
+}
+
+type RecordIteratorOption func(*recordIterator)
+
+func (r *recordIterator) Apply(opts ...RecordIteratorOption) {
+	for _, opt := range opts {
+		opt(r)
+	}
+}
+
+func RecordIteratorWithPage(config IteratorConfig) RecordIteratorOption {
+	return func(t *recordIterator) {
+		t.indexKey = config.key
+		t.validIndex = func(it columnIterator) bool {
+			if it == nil {
+				return false
+			}
+			// cache the type of Value()
+			switch it.Value().(type) {
+			case int64:
+				minStep := int64(config.min)
+				maxStep := int64(config.max)
+				t.validIndex = func(it columnIterator) bool {
+					step, ok := it.Value().(int64)
+					if !ok {
+						return false
+					}
+					return step >= minStep && step < maxStep
+				}
```

Comment:
Having `validIndex` replace itself is a tricky way to implement a stateful function. I'm not convinced this even improves performance, is it really needed? I'd keep it simple in this PR and optimize later.